File: data_structure_and_algorithm/hyejin/lecture02/mergeSort.js
const merge = (array1, array2) => {
  let mergedList = [];
  let i = 0; // 왼쪽 배열의 index
  let j = 0; // 오른쪽 배열의 index

  // 비교 후 append
  while (i < array1.length && j < array2.length) {
    if (array1[i] > array2[j]) {
      mergedList = mergedList.concat(array2[j]);
      j += 1;
    } else {
      mergedList = mergedList.concat(array1[i]);
      i += 1;
    }
  }

  // 기저 조건 : append 이후 다 더했을 때
  if (i === array1.length) mergedList = mergedList.concat(array2.slice(j));
  if (j === array2.length) mergedList = mergedList.concat(array1.slice(i));

  return mergedList;
};

const mergeSort = (array) => {
  // 기저 조건 : 1개일 때
  if (array.length <= 1) return array;

  let mid = array.length / 2;

  const leftHalf = array.slice(0, mid);
  const rightHalf = array.slice(mid);

  return merge(mergeSort(leftHalf), mergeSort(rightHalf));
};

console.log(mergeSort([1, 3, 5, 7, 9, 11, 13, 11]));
console.log(mergeSort([28, 13, 9, 30, 1, 48, 5, 7, 15]));

File: data_structure_and_algorithm/kim_kihyeon/lecture02/merge_sort.py
# 합병 정렬 Basic
# 1. Divide  : 리스트를 반으로 나눈다
# 2. Conquer : 왼쪽 리스트와 오른쪽 리스트를 각각 정렬한다
# 3. Combine : 정렬된 두 리스트를 하나의 정렬된 리스트로 합병한다


def merge(list1, list2):
    # list1, list2는 이미 정렬되어있음
    merged_list = []
    i = 0
    j = 0
    
    # list1과 list2를 돌면서 merged_list에 추가
    while i < len(list1) and j < len(list2):
        if list1[i] > list2[j]:
            merged_list.append(list2[j])
            j += 1
        else:
            merged_list.append(list1[i])
            i += 1
            
    # i가 len(list1)인 경우 list1이 소진된 것
    if  i == len(list1):
        merged_list.append(list2[j:])
    elif j == len(list2):
        merged_list.append(list1[i:])

    return merged_list


def merge_sort(my_list):

    # base case : 길이가 0 or 1인 경우 이미 정렬된 리스트
    if len(my_list) < 2:
        return my_list
    
    mid = len(my_list) //2
    
    left_half  = my_list[:mid]
    right_half = my_list[mid:] # mid+1로 착각
    
    # recursive case : merge의 파라미터로 넘겨주기
    return merge(merge_sort(left_half), merge_sort(right_half))
    
